### PR TITLE
Add Histopathology Generalist Training

### PR DESCRIPTION
**Pull request by anwai98**
(WIP) Add histopathology generalist training

### CODE REVIEW

**Comment by anwai98** [comment]
This looks good to go. Before putting this to review, it expects three PRs from `torch-em` to be merged (minor data-related developments) - [163](https://github.com/constantinpape/torch-em/pull/163), [164](https://github.com/constantinpape/torch-em/pull/164), [165](https://github.com/constantinpape/torch-em/pull/165). The fine-tuning is in progress. Will keep posted here.

**Comment by constantinpape** [comment]
I merged the corresponding `torch_em` PRs. I had a quick look here and this looks good.

**Comment by anwai98** [comment]
The trainings went through. This PR now looks good to go!

File: finetuning/generalists/training/histopathology/obtain_hp_datasets.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    train_ds, val_ds = data_util.random_split(ds, [1 - val_fraction, val_fraction], generator=generator)
+    return train_ds, val_ds
+
+
+class BCSSLabelTrafo:
+    def __init__(self, label_choices: Optional[List[int]] = None):
+        self.label_choices = label_choices
+
+    def __call__(self, labels: np.ndarray) -> np.ndarray:
+        """Returns the transformed bcss data labels (use-case for SAM)"""
```
Comment (constantinpape) [inline finetuning/generalists/training/histopathology/obtain_hp_datasets.py:38]: I think we should use connected components instead of simple relabeling here. This way disconnected pieces of the same semantic class would actually be split.

Comment (constantinpape) [reply finetuning/generalists/training/histopathology/obtain_hp_datasets.py:38]: Currently `label_padding_trafo` uses `label_consecutive_trafo` inside, you can add a switch to instead use `skimage.measure.label`.

**Review by constantinpape — COMMENTED** [review]
Looks good overall, just one minor comment.
(And you don't have to rerun training because of this, let's just change it so that it will be taken into account for any new training runs.)

File: finetuning/generalists/training/histopathology/obtain_hp_datasets.py (hunk 1 above)
Comment (anwai98) [reply finetuning/generalists/training/histopathology/obtain_hp_datasets.py:41]: Yes, that makes sense. Thanks Constantin! (fixed it just now in the latest commit)

Comment (constantinpape) [reply finetuning/generalists/training/histopathology/obtain_hp_datasets.py:41]: Now this is a bit confusing because one of the code paths applies padding and the other one doesn't...

File: finetuning/generalists/training/histopathology/obtain_hp_datasets.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+
+def get_concat_hp_datasets(path, patch_shape):
+    label_dtype = torch.int64
+    sampler = MinInstanceSampler(min_num_instances=5)
+
+    # make lizard dataset splits into fractions
+    lizard_ds = datasets.get_lizard_dataset(
+        path=os.path.join(path, "lizard"), patch_shape=patch_shape, sampler=sampler, label_dtype=label_dtype,
+        raw_transform=raw_padding_trafo, label_transform=label_padding_trafo
```
Comment (constantinpape) [inline finetuning/generalists/training/histopathology/obtain_hp_datasets.py:92]: Using the `raw_padding_trafo` like this will remove data normalization from the data loader. I am not sure if this leads to any problems.

File: finetuning/generalists/training/histopathology/obtain_hp_datasets.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    monusac_ds = datasets.get_monusac_dataset(
+        path=os.path.join(path, "monusac"), patch_shape=patch_shape, split="train", sampler=MinInstanceSampler(),
+        label_transform=label_consecutive_trafo, ndim=2, label_dtype=label_dtype
+    )
+    monusac_train_ds, monusac_val_ds = _get_train_val_split(ds=monusac_ds)
+
+    # out of three folds (sets of data) of provided data, we use two for training and 1 for validation
+    pannuke_train_ds = datasets.get_pannuke_dataset(
+        path=os.path.join(path, "pannuke"), patch_shape=(1, *patch_shape), sampler=sampler, folds=["fold_1", "fold_2"],
+        label_transform=label_padding_trafo, raw_transform=raw_padding_trafo, ndim=2, label_dtype=label_dtype
```
Comment (constantinpape) [inline finetuning/generalists/training/histopathology/obtain_hp_datasets.py:125]: See above.

**Comment by anwai98** [comment]
The training works with the most recent fixes now. This is good to go.

**Review by constantinpape — APPROVED** [review]
